#Trabalho prático # 5
##Introdução à Computação Numérica

###Tópico: Solução de Problemas de Valor Inicial (PVI)

---

Nome: Riquelme Freitas Gomes

DRE: 120032785


---

**Instruções:** 

**Preencha o notebook** com seus dados, seus códigos e **rode todos os códigos** 

**Inclua  um relatório, na parte final do notebook,** com suas conclusões, respostas, observações e comentários relacionados com os tópicos e questões abordadas nas tarefas.  

**A elaboração dos códigos pode ser realizada em duplas, mas o resto do trabalho é individual.**

*Se você fez o seu trabalho em dupla inclua essa informação no relatório. Além disso, se você  também recebeu ajuda de um(a) colega/amigo(a) ou consultou/usou informações da internet ou outras fontes, coloque referências à elas no seu relatório.*


**Entrega do trabalho:**

*Data de entrega: 09/janeiro/2023* 

---

## Tarefa #1:

 Desenvolver códigos, na forma de funções que implementem: 

a) O Método de Euler. 

b) O Método  de Euler aperfeiçoado (ou Método de Heun). 

c) O Método de Runge-Kutta clássico de quarta ordem.

In [1]:
# a) Função que implementa o Método de Euler

def Euler(f, cond_inicial, intervalo, N):
  t0 = intervalo[0]
  tf = intervalo[1]
  delta_t = (tf - t0) / N
  yn = cond_inicial

  for n in range(N+1):
    if(n != 0): 
      yn += delta_t * f(tn, yn)
      
    tn = t0 + (n * delta_t)

  return yn

In [2]:
# b) Função que implementa o Método de Euler aperfeiçoado (ou Método de Heun)

def Heun(f, cond_inicial, intervalo, N):
  t0 = intervalo[0]
  tf = intervalo[1]
  delta_t = (tf - t0) / N
  yn = cond_inicial

  for n in range(N+1):
    tn = t0 + (n * delta_t)

    if(n != 0): 
      k2 = delta_t * f(tn, yn + k1)
      yn += (k1 + k2) / 2

    k1 = delta_t * f(tn, yn)

  return yn

In [3]:
# c) Função que implementa o Método de Runge-Kutta clássico de quarta ordem

def Kutta(f, cond_inicial, intervalo, N):
  t0 = intervalo[0]
  tf = intervalo[1]
  delta_t = (tf - t0) / N
  yn = cond_inicial

  for n in range(N+1):
    tn = t0 + (n * delta_t)

    if(n != 0): 
      k4 = delta_t * f(tn, yn + k3)
      yn += (k1 + 2*k2 + 2*k3 + k4) / 6

    k1 = delta_t * f(tn, yn)
    k2 = delta_t * f(tn + (delta_t / 2), yn + (k1 / 2))
    k3 = delta_t * f(tn + (delta_t / 2), yn + (k2 / 2))

  return yn

## Tarefa #2: Paraquedista em queda livre

Utilize os seus códigos (da tarefa anterior, itens a), b) e c) ) para responder a seguinte questão.   
**Qual será a velocidade vertical atingida  por um paraquedista em queda livre (a velocidade inicial é nula), $25\,s$ após o salto?**


a) Considere que a taxa de variação instantânea da velocidade do paraquedista é dada pela EDO
$$\frac{dv}{dt}=g-\frac{c}{m} v, $$
onde $g=9.8\,m/s^2$ é a aceleração da gravidade, $m=75\,kg$ é a massa do paraquedista e $c=12.5\,kg/s$ é o coeficiente de arrasto correspondente ao paraquedas. 

Para cada método implementado determine um número de passos tal que o erro na aproximação da velocidade final $v(25)$ seja inferior a $10^{-2}$, se sabemos que o valor exato da velocidade  pode ser calculado pela fórmula
$$v(t) = \frac{gm}{c}\left( 1-e^{-(c/m)t}\right).$$

Dica: Isto poderá ser feito por tentativa e erro.

b) Considere uma outra lei para a taxa de variação instantânea da velocidade do paraquedista dada pela EDO
$$\frac{dv}{dt}=g-\frac{C(v)}{m} v, $$
onde a função empiríca $C(v)$ que caracteriza a resistência do paraquedas é dada por 
$$C(v) = c\left[1+\alpha \left(\frac{v}{v_{max}}\right)^\beta\right],$$
em que $c=12.5\,kg/s$, $\alpha=0.2$, $\beta =1.1$ e $v_{max} = 46\,m/s$.  (Observe que no caso $\alpha=0$ este modelo coincide com o apresentado no item a).)

Obtenha a velocidade final usando cada um dos métodos implementados considerando o mesmo número de passos definido no item a)).
 
c) Compare as velocidades obtidas nos itens a) e b). 

In [4]:
# a)
import math

g = 9.8  # aceleração da gravidade
m = 75   # massa do paraquedista
c = 12.5  # coeficiente de arrasto
t0, tf = 0, 25  # tempo inicial e tempo final
v0 = 0   # velocidade inicial

# Fórmula que calcula o valor exato da velocidade
def v(t): 
  return ((g * m) / c) * (1 - math.pow(math.e, -t*(c/m)))

def taxa_de_variacao(t, v):
  return g - (c / m)*v

cond_inicial = v0
intervalo = [t0, tf]
valor_exato = v(tf)
metodos = [Euler, Heun, Kutta]
num_passos, velocidades = [], []

for funcao in metodos:
  n = 1
  aproximacao = funcao(taxa_de_variacao, cond_inicial, intervalo, n)
  erro = abs(aproximacao - valor_exato)
  
  while(erro >= math.pow(10, -2)):
    n += 1
    aproximacao = funcao(taxa_de_variacao, cond_inicial, intervalo, n)
    erro = abs(aproximacao - valor_exato)
  
  num_passos.append(n)
  velocidades.append(aproximacao)
  print(f"{funcao}\nNumero de passos: {n}\nVelocidade aproximada: {aproximacao}")
  print(f"Velocidade exata: {valor_exato}\nErro absoluto: {erro}\n")

<function Euler at 0x7f77c58d4310>
Numero de passos: 790
Velocidade aproximada: 57.898370544133854
Velocidade exata: 57.88837340837825
Erro absoluto: 0.009997135755604347

<function Heun at 0x7f77c58d4280>
Numero de passos: 35
Velocidade aproximada: 57.87850926962627
Velocidade exata: 57.88837340837825
Erro absoluto: 0.009864138751979112

<function Kutta at 0x7f77c8b3a9d0>
Numero de passos: 7
Velocidade aproximada: 57.88180552525808
Velocidade exata: 57.88837340837825
Erro absoluto: 0.0065678831201694265



In [5]:
# b)
import math

alfa = 0.2
beta = 1.1
vmax = 46

# Função que caracteriza a resistência do paraquedas
def C(v):
  return c * (1 + (alfa * math.pow((v / vmax), beta)))

def taxa_de_variacao(t, v):
  return g - (C(v) / m)*v

velocidades2 = []
i = 0

for funcao in metodos:
  n = num_passos[i]
  i += 1
  aproximacao = funcao(taxa_de_variacao, cond_inicial, intervalo, n)
  erro = abs(aproximacao - valor_exato)
  velocidades2.append(aproximacao)
  print(f"{funcao}\nNumero de passos: {n}\nVelocidade final: {aproximacao}\n")

<function Euler at 0x7f77c58d4310>
Numero de passos: 790
Velocidade final: 48.37494162974708

<function Heun at 0x7f77c58d4280>
Numero de passos: 35
Velocidade final: 48.36749824086869

<function Kutta at 0x7f77c8b3a9d0>
Numero de passos: 7
Velocidade final: 48.364997612328395



In [7]:
# c)
print("Comparando as velocidades obtidas nos itens a) e b)\n")

for k in range(len(velocidades)):
  v1 = velocidades[k]
  v2 = velocidades2[k]
  print(metodos[k])
  print(f"Item a: {v1}\nItem b: {v2}")
  print(f"Módulo da diferença: {abs(v1 - v2)}\n")

Comparando as velocidades obtidas nos itens a) e b)

<function Euler at 0x7f77c58d4310>
Item a: 57.898370544133854
Item b: 48.37494162974708
Módulo da diferença: 9.523428914386777

<function Heun at 0x7f77c58d4280>
Item a: 57.87850926962627
Item b: 48.36749824086869
Módulo da diferença: 9.51101102875758

<function Kutta at 0x7f77c8b3a9d0>
Item a: 57.88180552525808
Item b: 48.364997612328395
Módulo da diferença: 9.516807912929686



## Tarefa #3: Dinâmica predador-presa

Para modelar a dinâmica de das populações de duas espécies que interagem: uma como predadora e a outra  como  presa, é usado o seguinte sistema de duas EDOs conhecido como equações de Lotka-Volterra:

$$
\left\{
\begin{aligned}
\frac{d x_1}{dt} &= a x_1-b x_1 x_2\\
\frac{d x_2}{dt} &= -c x_2 + d x_1 x_2
\end{aligned}
\right.
$$

onde $x_1$ e $x_2$ representam o número de presas e predadores, respectivamente, $t$ o tempo e $a$, $b$, $c$, $d$ são coeficientes constantes que caracterizam o crescimento e a interação das populações de presas e predadores.
 
Considere que a dinâmica das populações de alces e lobos de uma região pode ser modelada pelo sistema de Lotka-Volterra com os coeficientes $a=0.3$, $b=0.01111$, $c=0.2106$, $d=0.0002632$. Se inicialmente na região haviam $705$ alces e  $25$ lobos, qual será o tamanho dessas populações  após $20\, anos$?

Utilize o seu código do método de RK de $4^a$ ordem (tarefa #1, item c)) para responder a essa questão. Procure uma aproximação com erro estimado inferior a $0.05$. Explique. 

In [46]:
def Runge_Kutta(f1, f2, cond_inicial1, cond_inicial2, intervalo, N):
  t0 = intervalo[0]
  tf = intervalo[1]
  delta_t = (tf - t0) / N
  x1 = cond_inicial1
  x2 = cond_inicial2

  for n in range(N+1):
    tn = t0 + (n * delta_t)

    if(n != 0): 
      k4 = delta_t * f1(x2, x1 + k3)
      k4_x2 = delta_t * f2(x1, x2 + k3_x2)
      x1 += (k1 + 2*k2 + 2*k3 + k4) / 6
      x2 += (k1_x2 + 2*k2_x2 + 2*k3_x2 + k4_x2) / 6

    k1 = delta_t * f1(x2, x1)
    k2 = delta_t * f1(x2 + (delta_t / 2), x1 + (k1 / 2))
    k3 = delta_t * f1(x2 + (delta_t / 2), x1 + (k2 / 2))

    k1_x2 = delta_t * f2(x1, x2)
    k2_x2 = delta_t * f2(x1 + (delta_t / 2), x2 + (k1_x2 / 2))
    k3_x2 = delta_t * f2(x1 + (delta_t / 2), x2 + (k2_x2 / 2))

  return x1, x2

def dx1_dt(x2, x1):
  return a*x1 - b*x1*x2

def dx2_dt(x1, x2):
  return -c*x2 + d*x1*x2

a = 0.3
b = 0.01111
c = 0.2106
d = 0.0002632
x1i = 705
x2i = 25
intervalo = [0, 20]
N = 1 

x1, x2 = Runge_Kutta(dx1_dt, dx2_dt, x1i, x2i, intervalo, N)
x1f, x2f = Runge_Kutta(dx1_dt, dx2_dt, x1i, x2i, intervalo, N*2)

while(not (abs(x1 - x1f) < 0.05 and abs(x2 - x2f) < 0.05)):
  x1, x2 = Runge_Kutta(dx1_dt, dx2_dt, x1i, x2i, intervalo, N)
  x1f, x2f = Runge_Kutta(dx1_dt, dx2_dt, x1i, x2i, intervalo, N*2)
  N *= 2

print(f"Após {intervalo[1] - intervalo[0]} anos, essas populações terão aproximadamente...")
print(f"{x1f} alces e {x2f} lobos.")

Após 20 anos, essas populações terão aproximadamente...
701.1627245271219 alces e 28.931188329271187 lobos.


# **Relatório**

Nome: Riquelme Freitas Gomes \
DRE: 120032785

\
## **Trabalho prático #5**
Os códigos presentes neste documento são todos de minha autoria e para implementar alguns deles consultei o seguinte material: https://www.dropbox.com/s/eiozt07go8l1e6h/slides-Resolu%C3%A7%C3%A3o%20de%20PVIs.pdf?dl=0

\
## **Tarefa #1:**
Nessa tarefa, desenvolvi códigos, na forma de funções, que implementam:

### **a) O Método de Euler.**

### **b) O Método de Euler aperfeiçoado (ou Método de Heun).**

### **c) O Método de Runge-Kutta clássico de quarta ordem.**

\
## **Tarefa #2: Paraquedista em queda livre**
Nessa tarefa, utilizei os meus códigos (da tarefa anterior, itens a), b) e c) ) para determinar qual será a velocidade vertical atingida por um paraquedista em queda livre (a velocidade inicial é nula),  25s  após o salto.

\
### **a)**
Nesse item, para cada um dos métodos implementados anteriormente, determinei um número de passos tal que o erro na aproximação da velocidade final $v(25)$ seja inferior a $10^{-2}$. Encontrei esse número de passos por tentativa e erro, isto é, fui incrementando o número de passos até o valor do erro atingir o intervalo requisitado. Obtive os seguintes resultados para os Metódos de Euler, de Heun e de Runge-Kutta clássico de quarta ordem:

```
MÉTODO DE EULER
Numero de passos: 790
Velocidade aproximada: 57.898370544133854
Velocidade exata: 57.88837340837825
Erro absoluto: 0.009997135755604347

MÉTODO DE HEUN
Numero de passos: 35
Velocidade aproximada: 57.87850926962627
Velocidade exata: 57.88837340837825
Erro absoluto: 0.009864138751979112

MÉTODO DE RUNGE-KUTTA
Numero de passos: 7
Velocidade aproximada: 57.88180552525808
Velocidade exata: 57.88837340837825
Erro absoluto: 0.0065678831201694265
```

Observamos que o método de Euler atingiu o erro estipulado com 790 passos, o método de Heun atingiu com 35 passos e o método de Runge-Kutta atingiu com apenas 7 passos.

\
### **b)**
Nesse item, considerando o mesmo número de passos definido no item a): 790, 35 e 7 para os Métodos de Euler, Heun e Runge-Kutta respectivamente e considerando uma outra lei para a taxa de variação instantânea da velocidade do paraquedista, obtive a velocidade final usando cada um desses métodos. Atingi os seguintes resultados: 

```
MÉTODO DE EULER
Numero de passos: 790
Velocidade final: 48.37494162974708

MÉTODO DE HEUN
Numero de passos: 35
Velocidade final: 48.36749824086869

MÉTODO DE RUNGE-KUTTA
Numero de passos: 7
Velocidade final: 48.364997612328395
```

\
### **c)** 
Nesse item, comparei as velocidades obtidas nos itens a) e b), explicitando o valor do módulo da diferença entre elas. Esses foram os resultados:

```
MÉTODO DE EULER
Item a: 57.898370544133854
Item b: 48.37494162974708
Módulo da diferença: 9.523428914386777

MÉTODO DE HEUN
Item a: 57.87850926962627
Item b: 48.36749824086869
Módulo da diferença: 9.51101102875758

MÉTODO DE RUNGE-KUTTA
Item a: 57.88180552525808
Item b: 48.364997612328395
Módulo da diferença: 9.516807912929686
```

Pude observar que a diferença entre as velocidades foi bem grande para qualquer método utilizado. Além disso, as velocidades encontradas no item b) foram sempre bem menores que as encontradas no item a). O que, nesse contexto, é um resultado que faz sentido já que no item b) estamos considerando uma função empiríca $C(v)$ que caracteriza a resistência do paraquedas, portanto é esperado que essas velocidades sejam menores.

\
## **Tarefa #3: Dinâmica predador-presa**
Nessa tarefa, utilizei o meu código do método de RK de  4a  ordem (tarefa #1, item c) para determinar qual será o tamanho das populações de alces e lobos de uma região após 20 anos, procurando uma aproximação com erro estimado inferior a $0.05$.

Para conseguir essa resposta, adaptei minha função que implementa o método de Runge-Kutta para que ela pudesse receber duas funções e duas condições iniciais e, assim, pudesse calcular o número de alces e o de lobos ao mesmo tempo. Além disso, para satisfazer a condição do erro estimado inferior a $0.05$, comecei com uma entrada de $N = 1$ e fui dobrando este valor e comparando com os resultados anteriores (ou seja, os que receberam como entrada a metade do valor de $N$ atual). Meu código repetiu esse processo até que a diferença entre os valores atuais e os anteriores fosse inferior a $0.05$. Desse modo, o valor final de N que primeiro satisfez a condição foi de $32768$. Com isso, essa condição de parada garantiu um valor de $N$ suficientemente grande e uma aproximação com erro estimado inferior a $0.05$. 

Obtive o seguinte resultado para o tamanho dessas populações após 20 anos: 

```
Após 20 anos, essas populações terão aproximadamente...
701.1627245271219 alces e 28.931188329271187 lobos.
```

Dessa forma, meu código determinou que o número aproximado de alces será de $701.1627245271219$ e o número de lobos será de $28.931188329271187$.